### ISSUE TITLE
decreased performance after upgrade from 2.10.2 to 2.11.1

### ISSUE THREAD

**Issue by beanieboi** [puma/puma#661]
hey everybody,

yesterday i upgraded puma from `2.10.2` to `2.11.1` and saw an increase (around 10X) of the time spent in the request queue on Heroku.

![screen shot 2015-03-05 at 20 48 58](https://cloud.githubusercontent.com/assets/29251/6522089/69054de6-c3e0-11e4-8270-17fbafcabc58.png)

what is the best way to gather information to track this down? i'm not sure where i should start tracking this down. i'm happy to upgrade again for a few hours to collect some data.

best,
ben

**Comment by evanphx** [comment]
Nothing else changed, just changing the version of puma? What does your config look like?

**Comment by beanieboi** [comment]
to restore the performance we just downgraded puma. (see screenshot)
![screen shot 2015-03-06 at 16 19 46](https://cloud.githubusercontent.com/assets/29251/6527651/ad91e1d6-c41c-11e4-9b0d-e704741f356f.png)

our config/puma.rb

```
workers Integer(ENV['PUMA_WORKERS'] || 2)
threads Integer(ENV['PUMA_MIN_THREADS']  || 8), Integer(ENV['PUMA_MAX_THREADS'] || 12)

rackup      DefaultRackup
environment ENV['RACK_ENV'] || 'development'

if dir = ENV["BOXEN_SOCKET_DIR"]
  bind "unix://#{dir}/mothership"
else
  port ENV["PORT"] || 3000
end
```

the actual values from Heroku:

```
Loading production environment (Rails 4.1.9)
irb(main):001:0> ENV['PUMA_WORKERS']
=> "5"
irb(main):002:0> ENV['PUMA_MIN_THREADS']
=> "8"
irb(main):003:0> ENV['PUMA_MAX_THREADS'] 
=> "10"
irb(main):004:0> ENV['RACK_ENV']
=> "production"
irb(main):007:0> ENV["BOXEN_SOCKET_DIR"]
=> nil
irb(main):008:0> ENV["PORT"]
=> "47778"
```
and from the Procfile

```
web: bundle exec puma -C config/puma.rb
```

**Comment by evanphx** [comment]
Hm. The request queue... What exactly is that measuring? /cc @schneems

**Comment by beanieboi** [comment]
the official Heroku explanation is here: 
https://devcenter.heroku.com/articles/newrelic#interpreting-new-relic-request-queuing

**Comment by evanphx** [comment]
Ok, it's the time from the heroku router getting the request to when NewRelic picks it up.

Between 2.10.2 and 2.11.1 that are in the request path, but none seem like they'd cause a big problem.

Are you willing to try using puma from git to see if we can isolate the change that caused this? If so, I'll let you know when I've got a branch setup and we can coordinate. We'll probably want to coordinate in the Gitter room, so let's schedule when we can work on this for about an hour.

**Comment by schneems** [comment]
"Request queuing" in new relic is an inherently flawed measurement. They're comparing the clock of two different computers and only stop measuring once the new relic middleware is hit. It's not only measuring queuing time it also measures any rack middleware before the new relic middleware and the server request processing time. This is in addition to any time spent actually queuing the request.

**Comment by evanphx** [comment]
Is it possible that the request time leaped up because the change was deployed on a dyno with a drifted clock then? Seems like just deploying it again might tell us.

**Comment by beanieboi** [comment]
i’ll give it a shot.

**Comment by dnishimura** [comment]
@beanieboi Were you able to verify this?

**Comment by beanieboi** [comment]
tl;dr; 
sorry for the false alert.

i deployed the new puma version 30mins ago and there is _no_ increase in the request queuing. it looks like there happened something in NewRelic/Heroku/Middleware which caused the time spent in the request queue to spike.

i'll reopen in case something happens later today and looks related to that.

thanks everybody for the help!

**Comment by dnishimura** [comment]
Thanks. I downgraded to version 2.10.2 from 2.11.1 and didn't see a difference as well.
